casper/src/test/resources/PoSTest.rho

```diff
@@ -166,27 +179,52 @@ match (
               for (@user1PubKey, _, @user1Vault <- setupCh) {
                 @posVault!("balance", *initialPosBalanceCh) |
                 @user1Vault!("balance", *initialUser1BalanceCh) |
+                @PoS!("getRewards", *initialRewardsCh) |
                 for(@initialPosBalance <- initialPosBalanceCh;
-                    @initialUser1Balance <- initialUser1BalanceCh) {
+                    @initialUser1Balance <- initialUser1BalanceCh;
+                    @initialRewards <- initialRewardsCh) {
                   @PoS!("pay", 100, *retCh) |
                   for ( @(true, _) <- retCh) {
 
                     @posVault!("balance", *finalPosBalanceCh) |
                     @user1Vault!("balance", *finalUser1BalanceCh) |
+                    @PoS!("getRewards", *finalRewardsCh) |
                     for(@finalPosBalance <- finalPosBalanceCh;
-                        @finalUser1Balance <- finalUser1BalanceCh) {
-                      rhoSpec!("assertMany",
-                        [
-                          ((initialUser1Balance - finalUser1Balance, "==", 100), "the user account decreases"),
-                          ((finalPosBalance - initialPosBalance, "==", 100), "the pos account increases"),
-                        ],
-                        *ackCh
-                      )
+                        @finalUser1Balance <- finalUser1BalanceCh;
+                        @finalRewards <- finalRewardsCh) {
+                      @ListOps!("fold", finalRewards.toList(), {}, *computeDelta, *deltaRewardsCh) |
+                      contract computeDelta(@(pk, finalReward), @acc, resultCh) = {
+                        resultCh!(acc.set(pk, finalReward - initialRewards.get(pk)))
+                      } |
+                      for (@deltaRewards <- deltaRewardsCh) {
+                        @ListOps!("forall", deltaRewards, *isPositive, *allRewardsPositiveCh) |
+                        contract isPositive(@n, resultCh) = { resultCh!(n > 0)} |
+
+                        @ListOps!("fold", deltaRewards.toList(), 0, *sumDelta, *deltaSumCh) |
+                        contract sumDelta(@(_, delta), @acc, resultCh) = {
+                          resultCh!(acc + delta)
+                        } |
+
+                        rhoSpec!("assertMany",
+                          [
+                            ((initialUser1Balance - finalUser1Balance, "==", 100), "the user account decreases"),
+                            ((finalPosBalance - initialPosBalance, "==", 100), "the pos account increases"),
+                            ((finalRewards.size(), "==", initialRewards.size()), "rewards map size doesn't change"),
+                            //TODO use arbitrary precision numbers instead of long ints
+                            ((98, "== <-", *deltaSumCh), "the sum difference should equal the payment"),
+                            ((true, "== <-", *allRewardsPositiveCh), "the rewards can only grow")
+                          ],
```

Comment:
I'm having a really hard time reading such long and nested rholang, especially in tests.

I believe this could benefit from:
 - extracting a high-level assertion for checking all the interesting state in one contract call
 - leveraging the 'pass channels holding values instead of bare values' approach
   - this includes an 'act' wrapper method lifting the regular value-accepting tested method into channel-accepting one, see `transfer` in the test linked below

Both can be seen in https://github.com/rchain/rchain/blob/dev/casper/src/test/resources/RevVaultTest.rho#L129

Reply:
I'll try to refactor the tests in a new PR